### PR TITLE
Parse lockspace and resource names as bytes

### PR DESCRIPTION
**Pull request by aabawer**
PR outline:
1. set hardcoded ls_names and res_names as bytes in tests.
2. shut tests involving lockspace or resource names as bytes as xfail for py3.
3. add pybytes converter to sanlock.c 
4. go over API and apply parser per each lockspace/resource API call in a dedicated patch for easy review.
5. re-enable xfail tests for py3.

### CODE REVIEW

File: tests/python_test.py
Comment (nirs) [inline tests/python_test.py]: Makes sense. With these change we covered all inputs on python 2 and 3.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:311]: Looks good.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:399]: Looks good, but the parse_args test can be fixed now.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:454]: Looks good but the parse_args tests can be fixed now.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:512]: With this test can should be able to make test_write_lockspace_parse_args pass without xfail. This is the advantage of incremental development.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:728]: We should be able to make test_write_resource_parse_args pass without xfail now.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:783]: test_add_lockspace_parse_args should pass now without xfail.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:811]: If we fail here, we already set the exception...

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c]: But the code bellow this (hidden by github) is:

```
 771     if (rv == 0) {
 772         Py_RETURN_TRUE;
 773     } else if (rv == -ENOENT) {
 774         Py_RETURN_FALSE;
 775     } else if (rv == -EINPROGRESS) {
 776         Py_RETURN_NONE;
 777     }
 778 
 779     __set_exception(rv, "Sanlock lockspace inquire failure");
 780     return NULL;
 781 }
 ```

So we will set bogus sanlock excpetion with rv=-1, which will be translated to 
errno 1 (EPERM) in __set_exception(), so on the python side we will see:

    sanlock.SanlockException: (1, 'Sanlock lockspace inquire failure', 'Operation not permitted')

One way to avoid it is to use a sentinel value that cannot be returned by sanlock. sanlock returns either system errors (-1 - -128) or sanlock errors (-201 - -278) (src/sanlock_rv.h).

So we can do:

```
#define BINDING_ERROR -1000
...
int rv = BINDING_ERROR;
...
    if (rv == 0) {
        Py_RETURN_TRUE;
    } else if (rv == -ENOENT) {
        Py_RETURN_FALSE;
    } else if (rv == -EINPROGRESS) {
        Py_RETURN_NONE;
    } else if (rv == BINDING_ERROR) {
        return NULL;
    }

    __set_exception(rv, "Sanlock lockspace inquire failure");
    return NULL;
}
```

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:904]: test_rem_lockspace_parse_args should pass without xfail with this change.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:1018]: Looks good, but the relevant test should be modified.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:1095]: Looks good, relevant tests should be modified.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:1148]: Looks good, missing test change.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:1216]: Looks good, missing test change.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:1286]: Looks good, missing test change.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c]: Why not add proper cleanup here like in previous patches?

We can do:

```
    if (fd < 0) {
        __set_exception(fd, "Unable to register event fd");
       goto finally;
    }

finally:
    Py_XDECREF(lockspace);

    if (fd < 0)
        return NULL;

    return Py_BuildValue("i", fd);
```

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:1554]: Should use same finally pattern like previous patch.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c]: Should use same finally pattern as previous patch.

File: tests/python_test.py
Comment (nirs) [inline tests/python_test.py:56]: We don't need the pyparam() wrapper now, but we can keep it for symmetry if you like.

**Review by nirs — CHANGES_REQUESTED** [review]
I can take some of the first patches now.

Most patches are ready, but it would be nice to fix the xfail for parse_args tests related to each patch
to prove that the patch works. With this I can merge most patches. If you cannot find an easy way to enable the tests relevant to the patches, I can take all the patches together.

Some patches are missing proper cleanup, adding it should be easy and I don't see reason to skip it and have double cleanup code.

**Comment by nirs** [comment]
Here is a simple way to enable the relevant tests for every patch:
https://github.com/nirs/sanlock/commits/amit-py3-incremental-tests
- https://github.com/nirs/sanlock/commit/b26ebb3c3337a51a3e3df17537b44dda899cc85a
  (this code was missing a test - best add the tests before as separate patch)
- https://github.com/nirs/sanlock/commit/21285c8edc3d813f926bbad819593e1710c3ceb9
  (same, added missing test)
- https://github.com/nirs/sanlock/commit/1e2af7e708224ad151c16963c5e9a6f197727be1
  (1 line change to make the test pass)

What do you think?

File: python/sanlock.c
Comment (aabawer) [reply python/sanlock.c]: We can. But it will larger than actual cleanup to be done in this case. It will be only for lockspace. We can always add finaly label when time will come to add more cleanups.

File: python/sanlock.c
Comment (aabawer) [reply python/sanlock.c]: See prev comment

File: python/sanlock.c
Comment (aabawer) [reply python/sanlock.c]: See prev comment

File: python/sanlock.c
Comment (nirs) [reply python/sanlock.c]: Not sure what is "larger then actual cleanup". Why not keep the code consistent?

File: tests/python_test.py
Comment (aabawer) [reply tests/python_test.py:56]: i know

**Comment by aabawer** [comment]
1.will add the cleanup for the small ones as well.
2.will add parse args tests for init_{lockspace, resource} as part of their API commits 
3.i think this PR has matured to the level it can be merged entirely after 1 & 2 so i dont think we should waste time on enabling each test that can be enabled per every patch in the PR, but leave it to the final test patch as it is.

**Comment by aabawer** [comment]
Updated PR with fixes.

**Comment by nirs** [comment]
I think we can take this PR in the current form, but the next PRs should have working test with the code implementing the change.

**Comment by aabawer** [comment]
good news.
will try, some tests involve more than a single API and i'd like to change
one API at a time for sake of review.

**Comment by nirs** [comment]
The big flow tests should not be affected as since the default path can be unicode with ascii content, which is already supported.  We can add small tests like the tests for the lockspace/resource names to make it easy to fix the API in small steps.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:777]: The bracing changes are not needed. I'll fix this locally before pushing.

**Review by nirs — APPROVED** [review]

File: tests/python_test.py
Comment (nirs) [inline tests/python_test.py:587]: Missing empty line, I will fix it locally before I push.

File: python/sanlock.c
Comment (nirs) [inline python/sanlock.c:1554]: There is a leak here. Will fix locally before pushing.

**Comment by nirs** [comment]
I merge this into my py3 branch as 1ef37674d0465c4cf37c0892334ffd100f1ca870 with few changes.

**Comment by aabawer** [comment]
Thanks.
Good catch for the leak.

**Pull request closed without merging**
